File: backend/database/task_lock.js
/**
 * PostgreSQL advisory lock 封装 - Node.js 版
 * 用于调度器任务互斥
 */
const crypto = require('crypto');
const { pool } = require('./db');

function hashTaskName(name) {
    const hash = crypto.createHash('md5').update(name).digest('hex');
    return parseInt(hash.slice(0, 8), 16);
}

/**
 * 获取任务锁
 * @param {string} taskName - 任务名称
 * @returns {Promise<{client, key}>}
 */
async function acquireLock(taskName) {
    const key = hashTaskName(taskName);
    const client = await pool.connect();
    try {
        await client.query('SELECT pg_advisory_lock($1)', [key]);
        return { client, key, taskName };
    } catch (e) {
        client.release();
        throw e;
    }
}

/**
 * 释放任务锁
 * @param {{client, key, taskName}} lock - acquireLock 返回的对象
 */
async function releaseLock(lock) {
    try {
        await lock.client.query('SELECT pg_advisory_unlock($1)', [lock.key]);
    } finally {
        lock.client.release();
    }
}

/**
 * 带锁执行任务
 * @param {string} taskName - 任务名称
 * @param {Function} taskFn - 任务函数
 */
async function withLock(taskName, taskFn) {
    const lock = await acquireLock(taskName);
    try {
        return await taskFn();
    } finally {
        await releaseLock(lock);
    }
}

module.exports = { acquireLock, releaseLock, withLock };

File: backend/database/task_lock.py
#!/usr/bin/env python3
"""
PostgreSQL advisory lock 封装 - Python 版
用于调度器任务互斥
"""
import hashlib
import logging
from contextlib import contextmanager

logger = logging.getLogger(__name__)


def _hash_task_name(name: str) -> int:
    """将任务名映射为 32 位整数锁 key"""
    h = hashlib.md5(name.encode()).hexdigest()
    return int(h[:8], 16)


@contextmanager
def task_lock(task_name: str, conn_factory=None):
    """
    上下文管理器：获取 advisory lock，任务完成后释放

    用法:
        from database.task_lock import task_lock
        from core.db import get_conn

        with task_lock('article_sync', get_conn):
            run_sync_pipeline()
    """
    key = _hash_task_name(task_name)

    def _acquire_and_yield(conn_ctx):
        with conn_ctx as conn:
            cursor = conn.cursor()
            cursor.execute("SELECT pg_advisory_lock(%s)", (key,))
            logger.info(f"[Lock] 获取成功: {task_name} (key={key})")
            try:
                yield conn
            finally:
                cursor.execute("SELECT pg_advisory_unlock(%s)", (key,))
                logger.info(f"[Lock] 释放: {task_name} (key={key})")

    if conn_factory is None:
        from core.db import get_conn
        yield from _acquire_and_yield(get_conn)
    else:
        yield from _acquire_and_yield(conn_factory())
